contracts/VetoERC20Voting.sol

```diff
@@ -0,0 +1,181 @@
+    uint256 public vetoVotesThreshold;
+    IERC20Votes public votesToken;
+    IVetoGuard public vetoGuard;
+    mapping(bytes32 => uint256) public transactionVetoVotes;
+    mapping(address => mapping(bytes32 => bool)) public userHasVoted;
+
+  /// @notice Initializes the contract that can only be called once
+  /// @param _vetoVotesThreshold The number of votes required to veto a transaction
+  /// @param _votesToken The 
+  /// @param _vetoGuard The address of the VetoGuard contract
+  /// @param _accessControl The address of the access control contract
+    function initialize(
+        uint256 _vetoVotesThreshold,
+        address _votesToken,
+        address _vetoGuard,
+        address _accessControl
+    ) external initializer {
+        __initBase(_accessControl, msg.sender, "Veto ERC20 Voting");
+        vetoVotesThreshold = _vetoVotesThreshold;
+        votesToken = IERC20Votes(_votesToken);
+        vetoGuard = IVetoGuard(_vetoGuard);
+    }
+
+    /// @notice Allows the msg.sender to cast veto votes on the specified transaction
+    /// @param to Destination address.
+    /// @param value Ether value.
+    /// @param data Data payload.
+    /// @param operation Operation type.
+    /// @param safeTxGas Gas that should be used for the safe transaction.
+    /// @param baseGas Gas costs for that are independent of the transaction execution(e.g. base transaction fee, signature check, payment of the refund)
+    /// @param gasPrice Maximum gas price that should be used for this transaction.
+    /// @param gasToken Token address (or 0 if ETH) that is used for the payment.
+    /// @param refundReceiver Address of receiver of gas payment (or 0 if tx.origin).
+    function castVetoVote(
+        address to,
+        uint256 value,
+        bytes memory data,
+        Enum.Operation operation,
+        uint256 safeTxGas,
+        uint256 baseGas,
+        uint256 gasPrice,
+        address gasToken,
+        address payable refundReceiver
+    ) external {
+        // Get the transaction hash
+        bytes32 transactionHash = getTransactionHash(
+            to,
+            value,
+            data,
+            operation,
+            safeTxGas,
+            baseGas,
+            gasPrice,
+            gasToken,
+            refundReceiver
+        );
+
+        // Check that user has not yet voted
+        require(
+            !userHasVoted[msg.sender][transactionHash],
+            "User has already voted"
+        );
+
+        // Get the block number the transaction was queued on the VetoGuard
+        uint256 queuedBlockNumber = vetoGuard.getTransactionQueuedBlock(
+            to,
+            value,
+            data,
+            operation,
+            safeTxGas,
+            baseGas,
+            gasPrice,
+            gasToken,
+            refundReceiver
+        );
+
+        // Check that the transaction has been queued
+        require(queuedBlockNumber != 0, "Transaction has not yet been queued");
```

Comment:
I have it such that queuing is a required step before voting occurs because we need a block number to count user's available votes at. The block that the tx is queued at is the block number we're using.

I'm definitely open to changing this so user's could veto before queuing has occurred - we just need to think about which block number to use if that is the case. Any thoughts on that?

Reply:
Another question - should we create a function that allows a user to change their opinion? perhaps you place yes or no to veto? or is this confusing to the user? thoughts ?

Reply:
Yeah I do think it makes sense to add a function that can allow users to reverse a veto vote. 

Definitely some kind of mechanism that allows a tx to get "un-vetoed". I'd like to talk with the team as a whole on this one to get some more thoughts

Reply:
are there similar veto contracts that we've seen that allow unvetoing?  I'm kind of of the opinion that you shouldn't be able to undo your vote

Reply:
That makes sense - is there a reason @herbig for this? It works like this for voting on the governance module however, I somewhat view this as a tad different.

Reply:
that being said, I don't feel super strongly about it, if it works that way with the governance module we could just as well be consistent with that

Reply:
I don't feel strongly either. I defer to @tbwebb22 on this functionality.

Reply:
Cool, lets stick with the current voting model for now, where user's can't undo their veto.